front/lib/api/assistant/actions/browse.ts

```diff
@@ -0,0 +1,378 @@
+  Ok,
+} from "@dust-tt/types";
+import { isLeft } from "fp-ts/lib/Either";
+
+import { runActionStreamed } from "@app/lib/actions/server";
+import type { BaseActionRunParams } from "@app/lib/api/assistant/actions/types";
+import { BaseActionConfigurationServerRunner } from "@app/lib/api/assistant/actions/types";
+import type { Authenticator } from "@app/lib/auth";
+import { AgentBrowseAction } from "@app/lib/models/assistant/actions/browse";
+import logger from "@app/logger/logger";
+
+function isValidJSONArray(json: string): boolean {
+  try {
+    return Array.isArray(JSON.parse(json));
+  } catch (e) {
+    return false;
+  }
+}
+
+interface BrowseActionBlob {
+  id: ModelId; // AgentBrowseAction
+  agentMessageId: ModelId;
+  urls: string[];
+  output: BrowseActionOutputType | null;
+  functionCallId: string | null;
+  functionCallName: string | null;
+  step: number;
+}
+
+export class BrowseAction extends BaseAction {
+  readonly agentMessageId: ModelId;
+  readonly urls: string[];
+  readonly output: BrowseActionOutputType | null;
+  readonly functionCallId: string | null;
+  readonly functionCallName: string | null;
+  readonly step: number;
+  readonly type = "browse_action";
+
+  constructor(blob: BrowseActionBlob) {
+    super(blob.id, "browse_action");
+    this.agentMessageId = blob.agentMessageId;
+    this.urls = blob.urls;
+    this.output = blob.output;
+    this.functionCallId = blob.functionCallId;
+    this.functionCallName = blob.functionCallName;
+    this.step = blob.step;
+  }
+
+  renderForFunctionCall(): FunctionCallType {
+    return {
+      id: this.functionCallId ?? `call_${this.id.toString()}`,
+      name: this.functionCallName ?? "browse",
+      arguments: JSON.stringify({ urls: this.urls }),
+    };
+  }
+
+  renderForMultiActionsModel(): FunctionMessageTypeModel {
+    let content = "BROWSE OUTPUT:\n";
+    if (this.output === null) {
+      content += "The browse failed.\n";
+    } else {
+      content += `${JSON.stringify(this.output, null, 2)}\n`;
+    }
+
+    return {
+      role: "function" as const,
+      name: this.functionCallName ?? "browse",
+      function_call_id: this.functionCallId ?? `call_${this.id.toString()}`,
+      content,
+    };
+  }
+}
+
+/**
+ * Params generation.
+ */
+
+export class BrowseConfigurationServerRunner extends BaseActionConfigurationServerRunner<BrowseConfigurationType> {
+  async buildSpecification(
+    auth: Authenticator,
+    {
+      name,
+      description,
+    }: { name?: string | undefined; description?: string | undefined }
+  ): Promise<Result<AgentActionSpecification, Error>> {
+    const owner = auth.workspace();
+    if (!owner) {
+      throw new Error("Unexpected unauthenticated call to `runBrowseAction`");
+    }
+
+    return new Ok({
+      name: name ?? "browse",
+      description: description ?? "Get content of a page.",
+      inputs: [
+        {
+          name: "urls",
+          description: "List of urls to fetch, as a JSON array.",
```

Comment:
We should type this as a JSON array instead of a string that includes JSON which will be highly misleading and error prone for models to properly escape and format